frontend/packages/hardhat/contracts/EthernalChatIncentivized.sol

```diff
@@ -5,95 +5,136 @@ import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+		dataInfo.merkleRoot = newMerkleRoot;
+
+		emit CIDUpdated(msg.sender, cid);
+	}
 
-    // function getChallenge(address addr) public view returns ( uint64 index ) OnlyStorageProvider { // return index
-    // // require ..
-    //     index = 0 ; // random;
-    // }
+	/// @notice Add funds to an existing DataInfo by sending tokens and updating
+	/// @param newPrice The new price in tokens
+	/// @dev Make sure the DataInfo is non zero (has been already created)
+	function addFundsForStorage(uint256 newPrice) public {
+		DataInfo memory dataInfo = mapDataInfo[msg.sender];
+		require(dataInfo.cid != bytes32(0), "DataInfo not found");
+		paymentToken.transferFrom(msg.sender, address(this), newPrice);
+		dataInfo.allocatedTokens += newPrice;
+	}
 
+	/// @notice Retrieves the CID (Content Identifier) stored for the user.
+	/// @return The CID associated with the user.
+	/// @dev Passes when the user has a CID stored. This function is view only, so it doesn't modify the state of the contract
+	function getCID() public view returns (bytes32) {
+		DataInfo memory dataInfo = mapDataInfo[msg.sender];
+		require(dataInfo.cid != bytes32(0), "No CID found for this address");
+		return dataInfo.cid;
+	}
 
-    /// @notice Allows storage to get a partial amount of reward in token, if it hasn't been already taken in this amount of time 
-    function getStorageReward() public view returns (bytes32) {
-        // bytes32 cid = addressToCID[msg.sender];
-        // require( cid.length > 0, "No CID found for this address");
-        // // Check if the chunk size given correspond to the expected size
-        // return cid;
-    }
+	function getChallenge(
+		address addr
+	) public view OnlyStorageProvider(addr) returns (uint64 index) {
+		require(
+			mapDataInfo[addr].cid != bytes32(0),
+			"No DataInfo found for this address"
+		);
+		index = uint64(
+			uint256(
+				keccak256(abi.encodePacked(block.timestamp, block.prevrandao))
+			) % mapDataInfo[addr].numberOfChunks
+		);
+	}
 
-    /// @notice Take out all the rewards in Eth based on the amount of tokens the address holds
-    function withdrawRewards() public view returns (bytes32) {
-        // bytes32 cid = addressToCID[msg.sender];
-        // require( cid.length > 0, "No CID found for this address");
-        // return cid;
-    }
+	/// @notice Allows storage to get a partial amount of reward in token, if it hasn't been already taken in this amount of time
+	function getStorageReward() public view returns (uint256) {
+		uint256 timePassed = 24 hours;
+		require(
+			block.timestamp == block.timestamp + timePassed,
+			"Enough time hasn't passed"
+		);
+		require(
+			mapDataInfo[msg.sender].cid != bytes32(0),
+			"No Datainfo stored by this address"
+		);
+		DataInfo memory dataInfo = mapDataInfo[msg.sender];
+		return dataInfo.allocatedTokens;
+	}
 
+	/// @notice Take out all the rewards in Eth based on the amount of tokens the address holds
+	function withdrawRewards() public {
```

Comment:
See my comment on the previous function : so this should get the rewards accumulated by the  storageProvider and I think should only be take out after a certain amount of time (we can see how to define that)